R/writeNamedRegionToFile.R

```diff
@@ -1,51 +1,52 @@
-#
-# This program is free software: you can redistribute it and/or modify
-# it under the terms of the GNU General Public License as published by
-# the Free Software Foundation, either version 3 of the License, or
-# (at your option) any later version.
-#
-# This program is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU General Public License for more details.
-#
-# You should have received a copy of the GNU General Public License
-# along with this program.  If not, see <https://www.gnu.org/licenses/>.
-#
-#############################################################################
-
-#############################################################################
-#
-# Wrapper function to write data to an Excel file in one line.
-# 
-# Author: Thomas Themel, Mirai Solutions GmbH
-#
-#############################################################################
-
-
-writeNamedRegionToFile <- function(file, data, name, formula = NA, ..., 
-		styleAction = XLC$STYLE_ACTION.XLCONNECT,clearNamedRegions=FALSE) {
-
-  wb <- loadWorkbook(file, create = !file.exists(file))  
-  setStyleAction(wb, styleAction)
- 
-  # clear existing named regions
-  existingNames <- getDefinedNames(wb)
-  toClear <- intersect(name[clearNamedRegions], existingNames)
-  clearNamedRegion(wb, toClear)
-  
-  if(!is.na(formula)) {
-    sheets = extractSheetName(formula)
-    createSheet(wb, sheets[sheets != ""])
-    createName(wb, name, formula)
-  }
-  
-  writeNamedRegion(wb, data, name, ...)
-  
-  saveWorkbook(wb)
-  invisible(wb)  
-}
+#############################################################################
+#
+# XLConnect
+# Copyright (C) 2010-2021 Mirai Solutions GmbH
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+#############################################################################
+
+#############################################################################
+#
+# Wrapper function to write data to an Excel file in one line.
+# 
+# Author: Thomas Themel, Mirai Solutions GmbH
+#
+#############################################################################
+
+
+writeNamedRegionToFile <- function(file, data, name, formula = NA, ..., 
+		styleAction = XLC$STYLE_ACTION.XLCONNECT,clearNamedRegions=FALSE) {
+
```

Comment:
We need to make use of the possible `worksheetName` argument (retrieved as e.g. `list(...)$worksheetName` or added to the function signature explicitly):

- When handling `clearNamedRegions` (might require quite some refactoring)
- Passed to `createName()`

Reply:
@riccardoporreca I've made changes to enable this, but part of it seems like a hack (use of `sapply`). 

Without it, the following occurs, which I don't really understand:
```
Error in mapply(FUN = function (...)  : 
  zero-length inputs cannot be mixed with those of non-zero length
> traceback()
12: mapply(FUN = function (...) 
    .jrcall(x, name, ...), character(0), list(new("jobjRef", jobj = <pointer: (nil)>, 
        jclass = "java/lang/Object")), SIMPLIFY = FALSE)
11: do.call("mapply", args = c(FUN = f, args, SIMPLIFY = .simplify))
10: doTryCatch(return(expr), name, parentenv, handler)
9: tryCatchOne(expr, names, parentenv, handlers[[1L]])
8: tryCatchList(expr, classes, parentenv, handlers)
7: tryCatch(..., Throwable = function(e) {
       if (!is.jnull(e$jobj)) {
           sw <- new(J("java.io.StringWriter"))
           pw <- new(J("java.io.PrintWriter"), sw)
           e$jobj$printStackTrace(pw)
           options(java.stacktrace = sw$toString())
           stop(paste(class(e)[1], e$jobj$getMessage(), sep = " (Java): "), 
               call. = FALSE)
       }
       else stop("Undefined error occurred!")
   })
6: jTryCatch(do.call("mapply", args = c(FUN = f, args, SIMPLIFY = .simplify)))
5: xlcCall(object, "clearNamedRegion", name, worksheetScope %||% 
       .jnull(), .simplify = FALSE)
4: clearNamedRegion(wb, toClear, worksheetScope = worksheetScope)
3: clearNamedRegion(wb, toClear, worksheetScope = worksheetScope)
2: writeNamedRegionToFile(file, df, name, formula = paste(worksheet, 
       "A1", sep = "!"))
```

Reply:
@spoltier, need to look at this more in detail and give it some more thoughts